### ISSUE TITLE
Coroutines - Add rule that ensures that functions that return Flow are not marked as suspend

### ISSUE THREAD

**Issue by mhernand40** [detekt/detekt#3086]
## Expected Behavior of the rule
`Flows` are intended to be cold observable streams. The act of simply invoking a function that returns a `Flow`, should not have any side effects. Only once collection begins against the returned `Flow`, should work actually be done.

**Violating code:**
```Kotlin
suspend fun observeSignals(): Flow<Unit> {
    val pollingInterval = getPollingInterval() // Done outside of the flow block.
    return flow {
        while (true) {
            delay(pollingInterval)
            emit(Unit)
        }
    }
}

suspend fun getPollingInterval(): Long {
    // Return the polling interval from some repository
    // in a suspending manner.
}

coroutineScope.launch {
    val flow = observeSignals()
    delay(5_000L)
    // Let's say the polling interval from the repository was updated to a new value by now.
    flow.collect {
        // Old polling interval used from when observeSignals() was invoked above.
        println("Signal received")
    }
}
```

**Correct code:**
```Kotlin
fun observeSignals(): Flow<Unit> {
    return flow {
        val pollingInterval = getPollingInterval() // Moved into the flow builder block.
        while (true) {
            delay(pollingInterval)
            emit(Unit)
        }
    }
}

suspend fun getPollingInterval(): Long {
    // Return the polling interval from some repository
    // in a suspending manner.
}

coroutineScope.launch {
    val flow = observeSignals()
    delay(5_000L)
    // Let's say the polling interval from the repository was updated to a new value by now.
    flow.collect {
        // Latest polling interval upon collection start.
        println("Signal received")
    }
}
```

## Context
When a function returns a `Flow`, the caller should be able to invoke it without the consequence of the function itself doing any heavy work. The caller should be able to assume that the work begins once collections begins against the returned `Flow`. A function marked as `suspend` that also returns a `Flow` signals that the act of simply invoking it causes some kind of suspend or computational work irrespective of the returned `Flow` being collected.

**Comment by schalkms** [comment]
I particularly like this idea for a new rule.
May I ask you if you could lend us a helping hand here to bring this rule into detekt?

**Comment by mhernand40** [comment]
Sure, I'd be happy to. 🙂

**Comment by schalkms** [comment]
Thanks! This is highly appreciated.
It would be great to get this rule into detekt.
Feel free to reach out to us, if you have any questions. We try to help and answer them as good as we can. 🙂

**Comment by mhernand40** [comment]
Thank you. I appreciate the help and thank you for letting me contribute. 🙂 

I am hoping that perhaps @elizarov and/or @qwwdfsad can chime in and give an opinion on this proposed rule as well. 🙂

**Comment by elizarov** [comment]
You have my blessing. It is a good rule to have.

**Comment by qwwdfsad** [comment]
Good work!

**Comment by mhernand40** [comment]
Quick update. I have started with a TDD approach defining a suite of tests for different scenarios to take into account. A few things I would like for this rule to take into account:
1. Able to detect implementations of `Flow`, e.g., `StateFlow`, `MutableStateFlow`, etc.
2. Able to infer that return type is a `Flow` even if function is using an expression body without an explicit return type.
3. Able to distinguish between `kotlinx.coroutines.flow.Flow` from `com.some.other.type.of.Flow`.

**Examples:**
```Kotlin
import kotlinx.coroutines.flow.flowOf

// Should raise warning
suspend fun values() = flowOf(1, 2, 3)
```
```Kotlin
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow

// Should raise warning
suspend fun values(): StateFlow<Int> = MutableStateFlow(value = 1)
```
```Kotlin
import kotlinx.coroutines.flow.MutableStateFlow

// Should raise warning
suspend fun values() = MutableStateFlow(value = 1)
```
```Kotlin
import kotlinx.coroutines.flow.*

// Should raise warning
suspend fun values(): Flow<Int> = flowOf(1, 2, 3)
```
```Kotlin
import com.some.other.type.of.Flow

// Should NOT raise warning
suspend fun values(): Flow<Step> {
    TODO()
}
```
```Kotlin
// Should NOT raise warning
suspend fun values(): Flow<Int> {
    TODO()
}
```
```Kotlin
// Should raise warning
suspend fun values(): kotlinx.coroutines.flow.Flow<Int> {
    return kotlinx.coroutines.flow.flowOf(1, 2, 3)
}
```

Perhaps it is just me, but I am having to rely on debugger breakpoint watch expressions and IDE assistance in order to discover what is available from `kotlin-compiler-embeddable`. Is Detekt suited for the level of specificity that I am trying to achieve? Thanks in advance!

**Comment by BraisGabin** [comment]
Yes, you can archive all of those points using detekt. You can take some ideas from `ImplicitUnitReturnType`, for example.

**Comment by mhernand40** [comment]
Alright, PR up for review (https://github.com/detekt/detekt/pull/3098). Thanks again.

**Comment by mhernand40** [comment]
Closed by https://github.com/detekt/detekt/pull/3098
